hal/common/mbed_mem_trace.c

```diff
@@ -0,0 +1,127 @@
+#include "critical.h"
+#include "toolchain.h"
+
+/******************************************************************************
+ * Internal variables, functions and helpers
+ *****************************************************************************/
+
+/* The current tracing mask (see mbed_mem_trace.h for details). */
+static uint8_t mem_trace_mask = MBED_MEM_TRACE_DISABLE;
+/* The callback function that will be called after a traced memory operations finishes. */
+static mbed_mem_trace_cb_t mem_trace_cb;
+/* 'already_tracing' guards "trace inside trace" situations (for example, the implementation
+ * of realloc() might call malloc() internally, and since malloc() is also traced, this could
+ * result in two calls to the callback function instead of one. */
+static uint8_t already_tracing;
+
+/******************************************************************************
+ * Public interface
+ *****************************************************************************/
+
+void mbed_mem_trace_setup(uint8_t mask, mbed_mem_trace_cb_t cb) {
+    mem_trace_mask = mask;
+    mem_trace_cb = cb;
+}
+
+void* mbed_mem_trace_malloc(void *res, size_t size) {
+    uint8_t expected = 0;
+
+    if (mem_trace_cb && (mem_trace_mask & MBED_MEM_TRACE_MALLOC)) {
+        if (core_util_atomic_cas_u8(&already_tracing, &expected, 1)) {
+            mem_trace_cb(MBED_MEM_TRACE_MALLOC, res, MBED_CALLER_ADDR(), size);
+            already_tracing = 0;
+        }
+    }
+    return res;
+}
+
+void* mbed_mem_trace_realloc(void *res, void *ptr, size_t size) {
+    uint8_t expected = 0;
+
+    if (mem_trace_cb && (mem_trace_mask & MBED_MEM_TRACE_REALLOC)) {
+        if (core_util_atomic_cas_u8(&already_tracing, &expected, 1)) {
+            mem_trace_cb(MBED_MEM_TRACE_REALLOC, res, MBED_CALLER_ADDR(), ptr, size);
+            already_tracing = 0;
+        }
+    }
+    return res;
+}
+
+void* mbed_mem_trace_calloc(void *res, size_t num, size_t size) {
+    uint8_t expected = 0;
+
+    if (mem_trace_cb && (mem_trace_mask & MBED_MEM_TRACE_CALLOC)) {
+        if (core_util_atomic_cas_u8(&already_tracing, &expected, 1)) {
+            mem_trace_cb(MBED_MEM_TRACE_CALLOC, res, MBED_CALLER_ADDR(), num, size);
+            already_tracing = 0;
+        }
+    }
+    return res;
+}
+
+void mbed_mem_trace_free(void *ptr) {
+    uint8_t expected = 0;
+
+    if (mem_trace_cb && (mem_trace_mask & MBED_MEM_TRACE_FREE)) {
+        if (core_util_atomic_cas_u8(&already_tracing, &expected, 1)) {
+            mem_trace_cb(MBED_MEM_TRACE_FREE, NULL, MBED_CALLER_ADDR(), ptr);
+            already_tracing = 0;
+        }
+    }
+}
+
+void mbed_mem_default_callback(uint8_t op, void *res, void *caller, ...) {
+    va_list va;
+    size_t temp_s1, temp_s2;
+    void *temp_ptr;
+
+    va_start(va, caller);
+    switch(op) {
+        case MBED_MEM_TRACE_MALLOC:
+            temp_s1 = va_arg(va, size_t);
+            printf(MBED_MEM_DEFAULT_TRACER_PREFIX "m:%p;%p-%u\n", res, caller, temp_s1);
```

Comment:
After looking at mbed-trace, I'm a bit hesitant to use it here instead of `printf`. It looks like a very nice tracing library, but that's the problem: it's a library that adds to the flash size, has its own needs in terms of memory allocation and its own configuration. `printf` looks to be a good default solution, with minimal impact on setup and flash size.